### DISCUSSION TITLE
Langchain agent support

### DISCUSSION THREAD

**Discussion by leobenkel** [open-webui/pipelines#91] [General]
Can I connect this UI to Langchain agent and tooling ? 

Thank you

**Comment by MikeHago** [comment]
That would be a very cool feature :-)

**Comment by mynumu** [comment]
Me too! I want this…or an option to bring your own langchain code. Especially to change the RAG implementation

**Comment by alex-roc** [comment]
Hi, I have the same question as OP. Maybe we can connect it with LangServe/FastAPI?

**Comment by gradimirs** [comment]
Looking for this feature as well!

**Comment by ShaunBoughey** [comment]
Would be an amazing feature to see, would allow custom RAG implimentations to still have a nice UI!

**Comment by tcztzy** [comment]
Is there any idea to implement this feature? Is namespace package as plugin system a good idea?

**Reply by tcztzy** [reply]
I used dify recently, maybe we can implement it like dify did?

**Reply by tcztzy** [reply]
I have two implementation plans:

1. namespace package plugin system
2. `dify`-like implementation

Plan 1: We needs load third party with `open-webui.(apps|agents|tools)` at `main.py` start server. The third party package can be upload to pypi.org or just on github (pip can install github dependencies). And then we can treat the agent as a normal model or add tools to chat like ChatGPT's plugin.

Plan 2: We can add an openapi app as tool, and then design a page to create agent. All agent will store in database instead of code compare with plan 1.

We can implement both of them.

**Reply by aouicher** [reply]
ok for me

**Reply by silenceroom** [reply]
> I have two implementation plans:
> 
> 1. namespace package plugin system
> 2. `dify`-like implementation
> 
> Plan 1: We needs load third party with `open-webui.(apps|agents|tools)` at `main.py` start server. The third party package can be upload to pypi.org or just on github (pip can install github dependencies). And then we can treat the agent as a normal model or add tools to chat like ChatGPT's plugin.
> 
> Plan 2: We can add an openapi app as tool, and then design a page to create agent. All agent will store in database instead of code compare with plan 1.
> 
> We can implement both of them.

option 2 sounds like a great idea.

**Comment by jlcoo** [comment]
Me too! I want this…or an option to bring your own langchain code. Especially to change the RAG implementation

**Comment by jaaacki** [comment]
me too!!!!

**Comment by HorstA** [comment]
Me too! That would be almost vital for me ;-)

**Comment by PlebeiusGaragicus** [comment]
Came here also looking for this.

I'd like to be able to serve my LangGraph agents via LangServe and use OpenWebUI to call that endpoint.  I just started looking into this.

Also, if I could get per-user usage paid via micropayments this would become a my perfect solution!

**Comment by dannongruver** [comment]
Feels like a perfect solution. Langgraph also. 

Could be the middleware dev env for LLM app development. A dev builds this middle tier (RAG+Tools+Lang chain/DSPy) and tests via webUI frontend…and later have an OpenWebUI API to reuse this tested middleware for downstream apps. 

This is what I’m struggling with. I want to develop an LLM app and quickly test ideas for multi-agent implementation. Only option rn is python dev/test but no UI interface. So I have to build this type of testing ground UI.

But I guess that might change the name Open “WebUI” 😅 to Open “API + WebUI”

**Comment by PlebeiusGaragicus** [comment]
I have figured out a solution for this using Pipelines.  Just build a pipeline that calls your LangChain/Graph agent.  Instead of using LangServe, I just build an endpoint with FastAPI that invokes my graph.  Then, you can also host your own vector database and whatever else your agent will need.

**Reply by solargoe** [reply]
Have you published it for the general audience? That would be nice

**Reply by PlebeiusGaragicus** [reply]
Refer to this PR: https://github.com/open-webui/pipelines/pull/158/files

It's close enough to my solution.

Although, now that I'm using LangGraph Studio I'm going to have to rework this.  It seems LangGraph is moving towards the "assistants API" and I'm struggling to make it compatible with open webui at the moment.  Instead of calling an .invoke() or api/chat endpoint, you use "threads" and "runs."  There's probably an "orchestration layer" that Open WebUI needs in order to make it work well.

**Reply by pedromello** [reply]
> Refer to this PR: https://github.com/open-webui/pipelines/pull/158/files
> 
> It's close enough to my solution.
> 
> Although, now that I'm using LangGraph Studio I'm going to have to rework this. It seems LangGraph is moving towards the "assistants API" and I'm struggling to make it compatible with open webui at the moment. Instead of calling an .invoke() or api/chat endpoint, you use "threads" and "runs." There's probably an "orchestration layer" that Open WebUI needs in order to make it work well.

Thanks, @PlebeiusGaragicus, for sharing this code! 
I built on it to create my own version and wanted to share it with anyone interested. 

You guys can check it out here: [LangGraph Pipeline](https://github.com/pedromello/pipelines/tree/main/examples/pipelines/langgraph).

There's a README to guide you through the setup — it’s pretty straightforward! 🚀

**Comment by hasani114** [comment]
I tried this because it seemed pretty simple. Wasn't sure there was anything else to it? It works...

`"""
title: Langchain Pipeline
author: open-webui
date: 2024-05-30
version: 1.0
license: MIT
description: A pipeline for creating and managing conversational agents using the Langchain library.
requirements: langchain-core, langchain-openai
"""

from typing import List, Union, Generator, Iterator
from schemas import OpenAIChatMessage
from pydantic import BaseModel
from langchain_openai import ChatOpenAI


class Pipeline:
    class Valves(BaseModel):
        CHAT_MODEL: str = ""

    def __init__(self):
        # Optionally, you can set the id and name of the pipeline.
        # Best practice is to not specify the id so that it can be automatically inferred from the filename, so that users can install multiple versions of the same pipeline.
        # The identifier must be unique across all pipelines.
        # The identifier must be an alphanumeric string that can include underscores or hyphens. It cannot contain spaces, special characters, slashes, or backslashes.
        # self.id = "pipeline_example"

        # The name of the pipeline.
        self.name = "Langchain Pipeline"
        self.valves = self.Valves(**{"CHAT_MODEL": "gpt-4o"})

    async def on_startup(self):
        # This function is called when the server is started.
        print(f"on_startup:{__name__}")
        print(f"Model: {self.valves.CHAT_MODEL}")
        print(f"Valves: {self.valves}")
        pass

    async def on_shutdown(self):
        # This function is called when the server is stopped.
        print(f"on_shutdown:{__name__}")
        pass

    async def on_valves_updated(self):
        # This function is called when the valves are updated.
        pass

    async def inlet(self, body: dict, user: dict) -> dict:
        # This function is called before the OpenAI API request is made. You can modify the form data before it is sent to the OpenAI API.
        print(f"inlet:{__name__}")

        print(body)
        print(user)

        return body

    async def outlet(self, body: dict, user: dict) -> dict:
        # This function is called after the OpenAI API response is completed. You can modify the messages after they are received from the OpenAI API.
        print(f"outlet:{__name__}")

        print(body)
        print(user)

        return body

    def pipe(
        self, user_message: str, model_id: str, messages: List[dict], body: dict
    ) -> Union[str, Generator, Iterator]:
        # This is where you can add your custom pipelines like RAG.
        print(f"pipe:{__name__}")

        check_weather = {
            "name": "check_weather",
            "description": "Check the current weather conditions for a given location.",
            "parameters": {
                "type": "object",
                "properties": {
                    "location": {
                        "type": "string",
                        "description": "The city or location to check weather for",
                    },
                    "units": {
                        "type": "string",
                        "enum": ["metric", "imperial"],
                        "description": "The unit system for temperature (metric: Celsius, imperial: Fahrenheit)",
                    },
                },
                "required": ["location", "units"],
            },
        }
        # If you'd like to check for title generation, you can add the following check
        if body.get("title", False):
            print("Title Generation Request")

        model = ChatOpenAI(model=self.valves.CHAT_MODEL)
        model_with_tools = model.bind_tools([check_weather])
        print(model_with_tools.to_json())
        response = model_with_tools.stream(messages)

        # Yield each chunk from the response
        for chunk in response:
            yield chunk.content`

**Reply by tuniverspinner** [reply]
@hasani114 how to use this? sorry noob question

**Reply by PlebeiusGaragicus** [reply]
The above is a "pipeline" which you can use with OpenWebUI.  Just read the docs about how to get pipelines working, then you go into admin settings > pipelines and then import this module.  It should show up as a new model to talk to.

**Comment by bartonzzx** [comment]
#467 I think my example can help you with that :)

**Reply by PlebeiusGaragicus** [reply]
Yes - this is a wonderful PR - thanks very much!!
